worker/rhpv3.go

```diff
@@ -498,111 +442,131 @@ func readSectorCost(pt *rhpv3.HostPriceTable) (types.Currency, error) {
-	return *pt.pt, time.Now().Before(pt.expiry.Add(priceTableValidityLeeway))
+
+	return ongoing, pt.update
 }
 
-// Update updates a price table with the given host using the provided payment
-// function to pay for it.
-func (pts *priceTables) Update(ctx context.Context, payFn PriceTablePaymentFunc, siamuxAddr string, hk types.PublicKey) (rhpv3.HostPriceTable, error) {
-	// Fetch the price table to update.
-	pt := pts.priceTable(hk)
+func (p *priceTable) fetch(ctx context.Context, revision *types.FileContractRevision) (hpt hostdb.HostPriceTable, err error) {
+	// convenience variables
+	hk := p.hk
+	w := p.w
+	b := p.w.bus
 
-	// Check if there is some update going on already. If not, create one.
-	pt.mu.Lock()
-	ongoing := pt.ongoingUpdate
-	var performUpdate bool
-	if ongoing == nil {
-		ongoing = &priceTableUpdate{
-			done: make(chan struct{}),
-		}
-		pt.ongoingUpdate = ongoing
-		performUpdate = true
+	// grab the current price table
+	p.mu.Lock()
+	hpt = p.hpt
+	p.mu.Unlock()
+
+	// price table is valid, no update necessary, return early
+	priceTableUpdateLeeway := -time.Duration(frand.Intn(60)) * time.Second
+	if time.Now().Before(hpt.Expiry.Add(priceTableValidityLeeway).Add(priceTableUpdateLeeway)) {
+		return
 	}
-	pt.mu.Unlock()
 
-	// If this thread is not supposed to perform the update, just block and
-	// return the result.
-	if !performUpdate {
+	// price table is valid and update ongoing, return early
+	ongoing, update := p.ongoingUpdate()
+	if ongoing && time.Now().Before(hpt.Expiry.Add(priceTableValidityLeeway)) {
+		return
+	}
+
+	// price table is being updated, wait for the update
+	if ongoing {
 		select {
 		case <-ctx.Done():
-			return rhpv3.HostPriceTable{}, errors.New("timeout while blocking for pricetable update")
-		case <-ongoing.done:
-		}
-		if ongoing.err != nil {
-			return rhpv3.HostPriceTable{}, ongoing.err
-		} else {
-			return *ongoing.pt, nil
+			return hostdb.HostPriceTable{}, errors.New("timeout while blocking for pricetable update")
+		case <-update.done:
 		}
+		return update.hpt, update.err
 	}
 
-	// Update price table.
-	var hpt rhpv3.HostPriceTable
-	err := withTransportV3(ctx, siamuxAddr, hk, func(t *rhpv3.Transport) (err error) {
-		hpt, err = RPCPriceTable(t, payFn)
-		return err
-	})
+	// this thread is updating the price table
+	defer func() {
+		update.hpt = hpt
+		update.err = err
+		close(update.done)
 
-	pt.mu.Lock()
-	defer pt.mu.Unlock()
+		p.mu.Lock()
+		if err == nil {
+			p.hpt = hpt
+		}
+		p.update = nil
+		p.mu.Unlock()
+	}()
 
-	// On success we update the pt.
-	if err == nil {
-		ongoing.pt = &hpt
-		pt.pt = &hpt
-		pt.expiry = time.Now().Add(hpt.Validity)
+	// fetch the host info from the bus, it might have a new price table already
+	// in which case we can return without having to update it
+	host, err := b.Host(ctx, hk)
+	if err == nil && host.PriceTable != nil && time.Now().Before(host.PriceTable.Expiry.Add(priceTableValidityLeeway)) {
+		hpt = *host.PriceTable
+		return
+	}
```

Comment:
I get the intention here but it's not bulletproof. The chance is very high that we end up with multiple workers fetching a pricetable. 

The only sure way to avoid that is by moving the "ongoingUpdate" locking that we got here to the bus. So basically you do

- /bus/pricetable/ongoingudate (or whatever name)
- Then there are 2 results
  - A: The bus blocks for a bit and then returns the pricetable
  - B: The bus returns immediately, telling you that you need to fetch the pricetable

In the B case you'd then POST the pricetable to the bus after fetching it and you unblock the other workers.



While I feel this is essential for larger scale setups it's probably fine to skip it for this PR though since it's already quite involved.

Reply:
I didn't make it bulletproof on purpose to be honest. I wanted to avoid moving any locking over to the bus because currently every worker fetches its own price table anyway. I do want to maybe eventually measure how well this works. Saving workers from updating their price tables is only one of the upsides for me, having the price table on the host be updated frequently was the main driver for me. That,and bootstrapping workers quickly so we can immediately upscale/downscale... not that fetching a PT is a big deal.

Reply:
I don't mind moving the locking over to the bus side though, at all. Perhaps it is a good idea to do it in a separate PR since I don't think it's that big of a deal to have it on the worker side for the time being.